Mathlib/Data/Num/Basic.lean

```diff
@@ -0,0 +1,705 @@
+
+/-- The successor of a `PosNum`. -/
+def succ : PosNum → PosNum
+  | 1 => bit0 one
+  | bit1 n => bit0 (succ n)
+  | bit0 n => bit1 n
+#align pos_num.succ PosNum.succ
+
+/-- Returns a boolean for whether the `PosNum` is `one`. -/
+def isOne : PosNum → Bool
+  | 1 => true
+  | _ => false
+#align pos_num.is_one PosNum.isOne
+
+/-- Addition of two `PosNum`s. -/
+protected def add : PosNum → PosNum → PosNum
+  | 1, b => succ b
+  | a, 1 => succ a
+  | bit0 a, bit0 b => bit0 (PosNum.add a b)
+  | bit1 a, bit1 b => bit0 (succ (PosNum.add a b))
+  | bit0 a, bit1 b => bit1 (PosNum.add a b)
+  | bit1 a, bit0 b => bit1 (PosNum.add a b)
+#align pos_num.add PosNum.add
+
+instance : Add PosNum :=
+  ⟨PosNum.add⟩
+
+/-- The predecessor of a `PosNum` as a `Num`. -/
+def pred' : PosNum → Num
+  | 1 => 0
+  | bit0 n => Num.pos (Num.casesOn (pred' n) 1 bit1)
+  | bit1 n => Num.pos (bit0 n)
+#align pos_num.pred' PosNum.pred'
+
+/-- The predecessor of a `PosNum` as a `PosNum`. This means that `pred 1 = 1`. -/
+def pred (a : PosNum) : PosNum :=
+  Num.casesOn (pred' a) 1 id
+#align pos_num.pred PosNum.pred
+
+/-- The number of bits of a `PosNum`, as a `PosNum`. -/
+def size : PosNum → PosNum
+  | 1 => 1
+  | bit0 n => succ (size n)
+  | bit1 n => succ (size n)
+#align pos_num.size PosNum.size
+
+/-- The number of bits of a `PosNum`, as a `Nat`. -/
+def natSize : PosNum → Nat
+  | 1 => 1
+  | bit0 n => Nat.succ (natSize n)
+  | bit1 n => Nat.succ (natSize n)
+#align pos_num.nat_size PosNum.natSize
+
+/-- Multiplication of two `PosNum`s. -/
+protected def mul (a : PosNum) : PosNum → PosNum
+  | 1 => a
+  | bit0 b => bit0 (PosNum.mul a b)
+  | bit1 b => bit0 (PosNum.mul a b) + a
+#align pos_num.mul PosNum.mul
+
+instance : Mul PosNum :=
+  ⟨PosNum.mul⟩
+
+/-- `ofNatSucc n` is the `PosNum` corresponding to `n + 1`. -/
+def ofNatSucc : ℕ → PosNum
+  | 0 => 1
+  | Nat.succ n => succ (ofNatSucc n)
+#align pos_num.of_nat_succ PosNum.ofNatSucc
+
+/-- `ofNat n` is the `PosNum` corresponding to `n`, except for `ofNat 0 = 1`. -/
+def ofNat (n : ℕ) : PosNum :=
+  ofNatSucc (Nat.pred n)
+#align pos_num.of_nat PosNum.ofNat
+
+open Ordering
+
+/-- Ordering of `PosNum`s. -/
+def cmp : PosNum → PosNum → Ordering
+  | 1, 1 => eq
+  | _, 1 => gt
+  | 1, _ => lt
+  | bit0 a, bit0 b => cmp a b
+  | bit0 a, bit1 b => Ordering.casesOn (cmp a b) lt lt gt
+  | bit1 a, bit0 b => Ordering.casesOn (cmp a b) lt gt gt
+  | bit1 a, bit1 b => cmp a b
+#align pos_num.cmp PosNum.cmp
+
+instance : LT PosNum :=
+  ⟨fun a b => cmp a b = Ordering.lt⟩
+
+instance : LE PosNum :=
+  ⟨fun a b => ¬b < a⟩
+
+instance decidableLt : @DecidableRel PosNum (· < ·)
+  | a, b => by dsimp [LT.lt]; infer_instance
+#align pos_num.decidable_lt PosNum.decidableLt
+
+instance decidableLe : @DecidableRel PosNum (· ≤ ·)
+  | a, b => by dsimp [LE.le]; infer_instance
+#align pos_num.decidable_le PosNum.decidableLe
+
+end PosNum
+
+section
+
+variable {α : Type _} [One α] [Add α]
+
+set_option linter.deprecated false
```

Comment:
Not essential, but rather than turning options on and off you can write
```
set_option linter.deprecated false in
section
  ...
end
```

Reply:
or better yet just
```lean
section
set_option linter.deprecated false
  ...
end
```